### PR TITLE
Using the new MEF metadata to export DTBB targets

### PR DESCRIPTION
**Pull request by harleend**
###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/project-system/pull/9546)

### CODE REVIEW

**Comment by tmeschter** [comment]
@harleend Could you update the PR description to describe the problem this is addressing, and how it solves it?

**Comment by harleend** [comment]
The current DTBB design in CPS is subscription based. As new subscriptions are processed, it leads to the cancellation or requeuing of builds, which can sometimes result in builds being in an invalid state. To address this issue, we've recently defined new MEF metadata in CPS to pre-identify design time build targets to minimize the need for repeated build cancellations. This is also applicable to C# DevKit, where the logs indicate numerous cancellations.
 I have added/updated some of the rules definitions to use this new metadata. The subscriptions for these come in during solution load time.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/RuleExporter.cs [hunk 1]
```diff
@@ -27,6 +28,7 @@ private static class ProjectRules
             ///     for a project.
             /// </summary>
             [ExportRule(nameof(SupportedTargetFrameworkAlias), PropertyPageContexts.ProjectSubscriptionService)]
+            [ExportDesignTimeBuildTargets(nameof(SupportedTargetFrameworkAlias))]
```
Comment (lifengl) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/RuleExporter.cs:31]: I wonder both SupportedTargetFrameworkAlias & CollectedFrameworkReference might be only used in VS, but not VS Green...

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/RuleExporter.cs [hunk 2]
```diff
@@ -214,9 +231,56 @@ private static class OptionalWorkloadRules
             ///     Represents the evaluation items representing optional workloads.
             /// </summary>
             [ExportRule(nameof(SuggestedWorkload), PropertyPageContexts.ProjectSubscriptionService)]
+            [ExportDesignTimeBuildTargets(nameof(SuggestedWorkload))]
```
Comment (lifengl) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/RuleExporter.cs:234]: not sure about this one. It is not used in vs-green, but not sure it is always used in VS as well?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/RuleExporter.cs [hunk 3]
```diff
@@ -214,9 +231,56 @@ private static class OptionalWorkloadRules
...
             [AppliesTo(ProjectCapability.DotNet)]
             [Order(Order.Default)]
             public static int SuggestedWorkloadRule;
         }
+
+        /// <summary>
+        /// Contains rules applicable to the dependencies tree.
+        /// </summary>
+        private static class DependenciesTreeRules
```
Comment (lifengl) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/RuleExporter.cs:243]: I think we should not add dependencies tree rules in VS-Green.

**Review by lifengl — COMMENTED** [review]
My suggestion is that we might need disable some of them in VS green, because related code is not used or always used in VS Green.

**Comment by drewnoakes** [comment]
As discussed on Teams, I wonder if we can move these attributes to other types, so that we don't have to export rules via MEF in order to use this new CPS feature. We don't export/embed localised rules currently, and rules may also be provided dynamically via `IRuleProvider` (I think that's the interface).

**Comment by drewnoakes** [comment]
This was replaced by #9561. Closing.

**Pull request closed without merging**